torch/distributed/distributed_c10d.py

```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def _object_to_tensor(obj):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
+    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
+    local_size = torch.LongTensor([input_tensor.numel()])
+    return input_tensor, local_size
+
+
+def _tensor_to_object(tensor, tensor_size):
+    buf = io.BytesIO(tensor.numpy().tobytes()[:tensor_size])
+    buf.seek(0)
+    out = torch.load(buf)
+    return out
+
+
+def all_gather_object(object_list, obj, group=group.WORLD):
+    """
+    Gathers picklable objects from the whole group into a list. Similar to
+    all_gather, but Python objects can be passed in. Note that the object must
+    be pickable in order to be gathered.
+
+    Arguments:
+        object_list (list[Any]): Output list. It should be correctly sized as the
+        size of the group for this collective and will contain the output.
+        object (Any): Pickable Python object to be broadcast from current process.
+        group (ProcessGroup, optional): The process group to work on
+
+    Returns:
+        None. If the calling rank is part of this group, the output of the
+        collective will be populated into the input object_list.
+
+    .. note:: Note that this API differs slightly from the all_gather collective
+        since it does not provide an async_op handle and thus will be a blocking
+        call.
+    """
+    if _rank_not_in_group(group):
+        return
+
+    input_tensor, local_size = _object_to_tensor(obj)
+    # Gather all local sizes. This is so that we can find the max size, and index
+    # until the correct size when deserializing the tensors.
+    group_size = get_world_size(group=group)
+    object_size_list = [torch.LongTensor([0]) for _ in range(group_size)]
+    gloo_group = None
+    # Avoid copying intermediate tensors back and forth to CUDA by using gloo PG
+    # for object collectives.
+    if get_backend(group) == Backend.NCCL:
+        gloo_group = new_group(
```

Comment:
this will introduce one more rendezvous, will this be faster than copy the tensor to GPU when using NCCL backend?

Reply:
I see, maybe we can just go with the transfer to CUDA approach. I was thinking if we can cache this gloo group, then we might pay less amortized cost if this function is called many times.

Reply:
I see, let's use the provided group in this PR. If we got complain regarding its speed, we can find ways to optimize. Another reason for not creating a new Gloo group is that users might build `distributed` package without Glo.

Reply:
Sounds good, I think this can work for `all_gather_object` but for `gather_object`, we get the error `RuntimeError: ProcessGroupNCCL does not support gather`. I guess if users try to call `all_gather_object` with NCCL backend, it makes sense to throw, since NCCL does not implement gather. So should we just throw in this case?

Reply:
I see. Yep, I think it makes sense to just throw when calling `gather_object` with NCCL backend.